public/codenotes.php

```diff
@@ -14,8 +14,159 @@
 
 getCodeNotes($gameID, $codeNotes);
 
-RenderContentStart('Code Notes - ' . $gameData['Title']);
+RenderContentStart('Code Notes - ' . htmlspecialchars_decode($gameData['Title']));
 ?>
+<script>
+// If the user types some changes in for a note but cancels,
+// revert the input field's value to the note's original contents.
+/** @type Record<number, string> */
+const cachedInitialEditValues = {};
+
+/**
+ * Toggle the editing state of a code note row and 
+ * show/hide the elements necessary to perform an edit.
+ * 
+ * @param {HTMLTableRowElement} rowEl - The row element to toggle editing state for.
+ * @param {boolean} isEditing - Whether the row is transitioning to edit mode or not.
+ */
+function setRowEditingEnabled(rowEl, isEditing) {
+    const rowElementVisibilities = [
+        { className: 'edit-btn', showWhen: !isEditing },
+        { className: 'note-display', showWhen: !isEditing },
+        { className: 'note-edit', showWhen: isEditing },
+        { className: 'save-btn', showWhen: isEditing },
+        { className: 'cancel-btn', showWhen: isEditing },
+    ];
+
+    for (const visibility of rowElementVisibilities) {
+        const targetEl = rowEl.querySelector(`.${visibility.className}`);
+        
+        if (visibility.showWhen === true) {
+            targetEl.classList.remove('hidden');
+        } else {
+            targetEl.classList.add('hidden');
+        }
+    }
+}
+
+/**
+ * Enable edit mode for a specific row.
+ * 
+ * @param {number} rowIndex - The index of the row to enable edit mode for.
+ */
+function beginEditMode(rowIndex) {
+    const rowEl = document.getElementById(`row-${rowIndex}`);
+    setRowEditingEnabled(rowEl, true);
+
+    // In the event the user cancels their edit, we'll restore whatever
+    // the original note value is to the input field. We use this cache
+    // in order to do so.
+    const noteEditEl = rowEl.querySelector('.note-edit');
+    cachedInitialEditValues[rowIndex] = noteEditEl.value;
+
+    // Calculate the number of rows based on the number of newline characters.
+    const rowCount = (noteEditEl.value.match(/\n/g) || []).length + 1;
+
+    // Set the textarea rows attribute, ensuring a minimum of 2 rows.
+    noteEditEl.rows = Math.max(rowCount, 2);
+}
+
+/**
+ * Cancel edit mode for a specific row.
+ * 
+ * @param {number} rowIndex - The index of the row to cancel edit mode for.
+ */
+function cancelEditMode(rowIndex) {
+    const rowEl = document.getElementById(`row-${rowIndex}`);
+
+    // Restore the cached value so unsaved edits are not persisted.
+    const noteEditEl = rowEl.querySelector('.note-edit');
+    noteEditEl.value = cachedInitialEditValues[rowIndex];
```

Comment:
This seems to work, but I'm wondering how it could be merged with the `isSavedNoteDifferent` logic. Should that check the `cachedInitialEditValue`, or can `cachedInitialEditValues` be eliminated in favor of the `noteDisplayEl.innerHTML`?

Reply:
I like the idea of getting rid of `cachedInitialEditValues` and querying for the current `noteDisplayEl.innerHTML` value. This revision has been made in https://github.com/RetroAchievements/RAWeb/pull/1440/commits/fe725a1a83c1e1a6c8999f51722123c944639e84.